bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LastAddForcedCursor.java

```diff
@@ -0,0 +1,145 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.bookie;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import org.apache.bookkeeper.client.api.WriteAdvHandle;
+import org.apache.bookkeeper.proto.BookieProtocol;
+
+/**
+ * Handle the LastAddForced entryId. As entries do not arrive to the Journal in a known order we have to handle
+ * increments of this pointer. This is how it works: we keep a value for the current {@link #lastAddForced} entryId and
+ * a list of continuous ranges of ids which are known to have been persisted but are 'after' {@link #lastAddForced}.
+ * <p>
+ * Examples:
+ * <ul>
+ * <li>0, 1, 2, 3 we have maxAddForced = 3 (all entries from 0 to 3 are stored)</li>
+ * <li>0, 1, 3, 5 we have maxAddForced = 1 (all entries from 0 to 1 are stored, 2 and 4 are not stored)</li>
+ * <li>1, 2, 3 we have maxAddForced = -1 (not available, as 0 is missing, so there is an gap)</li>
```

Comment:
There can never be contiguous entries in the bookie perspective if Ensemble Size > WQ

Reply:
Currently in BP14 we are restricting to ES = WQ >= AQ.
So eventually there will be no hole, only temporary ones due to WriteAdvHandle  or temporary network errors/retransmission.

Reply:
Hmm, So this is not just a test statement, this implementation itself is going to restrict the ensemble size. usually, people select this mode for better performance, limiting this reduces the striping and hence throughput. What other options do we have? so that the solution doesn't push us to a corner?

Reply:
@jvrao from the very beginning of BP14 we decided not to implement "striping" in the first implement because it brings many problems.

In my first proposal I had a "force" API in which the client was sending not just the 'last written entry id' but a range of entries, in order to ask the bookie about the segment of ledger which is the current 'tail segment', because with striping the bookie  is not receiving all the entries of the ledger.

As an initial implementation it is fine to have ES = WQ, we can just enhance this feature in the future. My initial idea was to make the client send the range of entries which defines the current segment (bookie do not have info about location of entries). This way the bookie can report to the client the actual status of the entries in the range and the client is able to advance LAC.

Client will advance LAC only in presence of the guarantee of sequence of entries fully persisted  on AQ bookies, without holes.
In other words:
*the range 0...LAC is always persisted durably on AQ bookies*

the force() API enables the client to wait for an fsync on all the bookies on the current ensemble.
In presence of ES > WQ entries will be striped so the force API() is far more complex, not infeasible, but really complex, because the client must ask to the bookies for each segment.

Ok for talking about this on next meeting.

cc @sijie

Reply:
I think we are mixing up multiple things here. Let me see if I can clarify
1. Client advances LAC is based on Bookie response on write. 
2. Bookie decides to persist the write or not based on the addentry flag.

If we split them, then there is NO change needed on the LAC logic. 
No need to maintain this cursor or track lastAddCommit.

The application is taking the responsibility of potentially losing data between sync points.
And just like in any other storage subsystem, if a writer is acknowledged successfully
a reader must be allowed to read that.

So, one guarantee we break in this model - In this durability model - "If an entry is read once you can always read it". That is expected.

Also on the sync, we don't need to return *any* entryId (last commit id or something); A successful sync guarantees that all the data written on that extent is synced. 

Hope this clarifies and makes the patch simple. I briefly mentioned to @sijie  too, and he did not raise any major concern with this model.

Reply:
Extending my thought...
This would make the recovery complicated if we update the piggybacked LAC too.
Because LAC may point to something that is not there.

So, we can have nonDurableLAC on the client side. This gets updated just like today's LAC, and LAC = nonDuarbleLAC on successful SYNC. With this model, you can achieve what you are trying to do now without maintaining the cursor. But this won't allow reads on all entries that were written after the previous sync.